### PR TITLE
Generalize network_interface table to allow for new kinds of NICs.

### PR DESCRIPTION
**Pull request by luqmana**
Since I wanted to split up #2419 a bit, took the opportunity to get rid of the
duplicate table I added to get going initially.

Previously, all NIC records in the DB were tied to a guest instance. In
enabling OPTE usage for services, it'd be nice to be able to reuse a lot
of the same NetworkInterface machinery we already have without
duplicating it completely. This commit adds a new `kind` column to the
`network_interface` table which at the moment will be either 'instance'
(NICs attached to a guest VM and exposed in the external API) or
'service' (NIC associated with an internal control plane service). The
previous `instance_id` column is renamed to `parent_id` and the table to
which it refers to as a FK is now dependant on the kind (either
`instance` or `service`).

Since a lot of the db and authz lookup macros end up relying on the
specific column name, this also introduces database views for each kind
which can be queried as if they were their own tables. This also allows
differentiating the different kinds of NICs as necessary.

CockroachDB, unlike Postgres, does not allow inserting or updating into
simple view and so we also model the base table itself to execute
queries that modify it.

From an external perspective, this doesn't change anything (modulo some
renaming from `NetworkInterface*` -> `InstanceNetworkInterface*`) in
that all external APIs still only deal with instance NICs.

There are some basic definitions for service NICs included but those
will be more fleshed in subsequent commits.

### CODE REVIEW

File: nexus/db-model/src/network_interface.rs [hunk 1]
```diff
@@ -14,36 +17,184 @@ use nexus_types::identity::Resource;
...
+    pub kind: NetworkInterfaceKind,
+    pub parent_id: Uuid,
+
     pub vpc_id: Uuid,
     pub subnet_id: Uuid,
+
     pub mac: MacAddr,
     // TODO-correctness: We need to split this into an optional V4 and optional V6 address, at
-    // least one of which will always be specified.
+    // least one of which will lways be specified.
```
Comment (zephraph) [inline nexus/db-model/src/network_interface.rs:48]:
```suggestion
    // least one of which will always be specified.
```

**Comment by zephraph** [comment]
This is great. I was working on implementing silo level images and doing it in a way that was much harder than what you've done here. I'm definitely interested in following this approach.

File: nexus/db-model/src/network_interface.rs [hunk 2]
```diff
@@ -14,36 +17,184 @@ use nexus_types::identity::Resource;
...
+    #[diesel(column_name = is_primary)]
+    pub primary: bool,
+}
+
+impl NetworkInterface {
+    /// Treat this `NetworkInterface` as an `InstanceNetworkInterface`.
+    ///
+    /// # Panics
+    /// Panics if this isn't an 'instance' kind network interface.
+    pub fn as_instance(self) -> InstanceNetworkInterface {
```
Comment (zephraph) [inline nexus/db-model/src/network_interface.rs:110]: Is there a reason not to do a `TryFrom` instead?

Comment (luqmana) [reply nexus/db-model/src/network_interface.rs:110]: Hmm, mostly just cause the only places I used it expect the conversion to be infallible and so using `TryFrom`/`TryInto` would just mean shuffling the panic/unwrap.

**Review by jmpesp — APPROVED** [review]
Looks good, one comment needs fixing

**Pull request merged**